src/idl/tests/annotation.c

```diff
@@ -455,93 +455,152 @@ CU_Test(idl_annotation, redefinition)
-      const idl_struct_t *s = (const idl_struct_t *)pstate->root;
-      CU_ASSERT(idl_is_struct(s));
-      const idl_member_t *m = s->members;
-      CU_ASSERT(idl_is_member(m));
-      CU_ASSERT_PTR_NOT_NULL(m->id.annotation);
-      CU_ASSERT_EQUAL(m->id.value, tests[i].h);
-    }
-    idl_delete_pstate(pstate);
-  }
+  for (size_t i = 0; i < sizeof(tests)/sizeof(id_test_t); i++)
+    test_id(tests[i]);
 }
 
-// x. do not allow annotation_appl in annotation
+CU_Test(idl_annotation, autoid_struct)
+{
+  static const id_test_t tests[] = {
+    {"struct s { char c; char d; };",                                               IDL_RETCODE_OK,             {IDL_SEQUENTIAL}, {0},            {0, 1}},                                  //implicit sequential autoid
+    {"struct s { @id(456) char c; char d; };",                                      IDL_RETCODE_OK,             {IDL_SEQUENTIAL}, {true, false},  {456, 457}},                              //implicit sequential autoid, starting from a specific id
+    {"@autoid struct s { char c; char d; };",                                       IDL_RETCODE_OK,             {IDL_HASH},       {0},            {0x00088a4au, 0x01e07782u}},              //implicit hash autoid
+    {"@autoid(HASH) struct s { char c, d; char e; };",                              IDL_RETCODE_OK,             {IDL_HASH},       {0},            {0x00088a4au, 0x01e07782u, 0x071767e1u}}, //explicit hash autoid
+    {"@autoid(SEQUENTIAL) struct s { char c, d; char e; };",                        IDL_RETCODE_OK,             {IDL_SEQUENTIAL}, {0},            {0, 1, 2}},                               //explicit sequential autoid
+    {"@autoid(SEQUENTIAL) struct s { @hashid char c; char d; };",                   IDL_RETCODE_OK,             {IDL_SEQUENTIAL}, {true, false},  {0x00088a4au, 0x00088a4bu}},              //explicit sequential autoid, with implicit hashid
+    {"@autoid(HASH) struct s { @id(123) char c; char d; };",                        IDL_RETCODE_OK,             {IDL_HASH},       {true, false},  {123, 0x01e07782u}},                      //explicit hash autoid, with explicit id
+    {"@autoid @autoid(SEQUENTIAL) struct s { char c; };",                           IDL_RETCODE_SEMANTIC_ERROR, {0},              {0},            {0}},                                     //conflicting duplicate autoid
+    {"@autoid(SEQUENTIAL) struct s { @id(456) char c; char d; @id(457) char e; };", IDL_RETCODE_SEMANTIC_ERROR, {0},              {0},            {0}},                                     //clashing sequential id fields
+    {"@autoid(HASH) struct s { char c; @hashid(\"c\") char d; };",                  IDL_RETCODE_SEMANTIC_ERROR, {0},              {0},            {0}}                                      //clashing hashid fields
+  };
 
-#if 0
-CU _ Test(idl_annotation, autoid_struct)
+  for (size_t i = 0; i < sizeof(tests)/sizeof(id_test_t); i++)
+    test_id(tests[i]);
+}
+
+CU_Test(idl_annotation, autoid_inheritance)
```

Comment:
I think it would be good to cover some weird scenarios here. So a derived of a derived that has a higher id than the intermediate. That sort of thing.

Reply:
I will make additional tests